rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala

```diff
@@ -0,0 +1,317 @@
+          val existingHead :: existingTail = affixBytes.drop(prefixPath.size)
+          val (dividedElems, _) = divideSkip(
+            incomingTail,
+            incomingHead,
+            newLeaf,
+            existingTail,
+            existingHead,
+            existingPointer,
+            prefixPath
+          )
+          Applicative[F].pure(dividedElems)
+
+        case TriePath(
+            elems :+ (pointerBlock: PointerBlock),
+            Some(Skip(affix, existingPointer)),
+            common
+            ) => // split existing skip
+          val incomingPath                 = remainingPath.drop(common.size)
+          val affixBytes                   = affix.toSeq.toList
+          val prefixPath                   = commonPrefix(incomingPath, affixBytes)
+          val incomingHead :: incomingTail = incomingPath.drop(prefixPath.size)
+          val existingHead :: existingTail = affixBytes.drop(prefixPath.size)
+          val (dividedElems, dividedTopPtr) = divideSkip(
+            incomingTail,
+            incomingHead,
+            newLeaf,
+            existingTail,
+            existingHead,
+            existingPointer,
+            prefixPath
+          )
+          val updatedPointerBlock = pointerBlock.updated((toInt(common.last), dividedTopPtr) :: Nil)
+          for {
+            result <- rehash(common.init, elems, updatedPointerBlock, dividedElems)
+          } yield result
+
+        case TriePath(elems :+ (pastPb: PointerBlock), None, common) => // add to existing node
+          val key          = remainingPath.drop(common.size)
+          val (ptr, nodes) = skip(key, newLeaf)
+          val updatedPb    = pastPb.updated((toInt(common.last), ptr) :: Nil)
+          for {
+            result <- rehash(common.init, elems, updatedPb, nodes.toList)
+          } yield result
+
+        case TriePath(elems :+ Skip(affix, LeafPointer(_)), _, path) => // update value
+          rehash(path.dropRight(affix.size.toInt), elems, Skip(affix, newLeaf), Nil)
+        case _ => Sync[F].raiseError(MalformedTrieError)
+      }
+
+    private def rehash(
+        path: KeyPath,
+        rest: Vector[Trie],
+        lastSeen: Trie,
+        acc: List[Trie]
+    ): F[List[Trie]] = {
+      type Params = (KeyPath, Vector[Trie], Trie, List[Trie])
+      def go(params: Params): F[Either[Params, List[Trie]]] =
+        params match {
+          case (
+              remainingPath: KeyPath,
+              remainingPrefixNodes: Vector[Trie],
+              nextLastSeen: Trie,
+              currentAcc: List[Trie]
+              ) =>
+            (remainingPrefixNodes, nextLastSeen) match {
+              case (head :+ (pointerBlock: PointerBlock), other) =>
+                val ptr     = pointer(other)
+                val updated = pointerBlock.updated((toInt(remainingPath.last), ptr) :: Nil)
+                Applicative[F].pure(
+                  (remainingPath.init, head, updated, currentAcc :+ nextLastSeen).asLeft
```

Comment:
You have quite a few usages of :+ which makes me worry about the complexity of the algorithm.

What's the complexity of rehash? To me it looks quadratic in the length of KeyPath. How big can the KeyPath be? Is this a concern?

Reply:
the complexity of the rehash method isn't the focus of this pr (it did not change since last implementation)

having said that: https://rchain.atlassian.net/projects/RCHAIN/issues/RCHAIN-3508?